test/jdk/java/awt/TextComponent/SelectionAndCaretColor.java

```diff
@@ -0,0 +1,157 @@
+    static Robot r;
+    static Frame frame;
+    static volatile int flips;
+
+    public static void main(String[] args) throws Exception {
+        try {
+            frame = new Frame("Selection and Caret color test");
+            r = new Robot();
+
+            EventQueue.invokeAndWait(() -> {
+                frame.setLayout(new BorderLayout());
+                tf.setFont(new Font("Monospaced", Font.PLAIN, 15));
+                ta.setFont(new Font("Monospaced", Font.PLAIN, 15));
+
+                frame.add(tf, BorderLayout.NORTH);
+                frame.add(ta, BorderLayout.SOUTH);
+                frame.setSize(200, 200);
+                frame.setVisible(true);
+            });
+            r.waitForIdle();
+            r.delay(1000);
+            test(tf);
+            test(ta);
+        } finally {
+            EventQueue.invokeAndWait(() -> {
+                if (frame != null) {
+                    frame.dispose();
+                }
+            });
+        }
+    }
+
+    private static int countFlips(TextComponent tc) {
+        int y = tc.getLocationOnScreen().y + tc.getHeight() / 2;
+        int x1 = tc.getLocationOnScreen().x + 5;
+        int x2 = tc.getLocationOnScreen().x + tc.getWidth() - 5;
+
+        int[] fb = {tc.getBackground().getRGB(), tc.getForeground().getRGB()};
+        int i = 0;
+        int flips = 0;
+
+        BufferedImage img = r.createScreenCapture(new Rectangle(x1, y, x2 - x1, 1));
+        for (int x = 0; x < x2 - x1; x++) {
+            int c = img.getRGB(x, 0);
+            if (c == fb[i]) {
+                ;
+            } else if (c == fb[1 - i]) {
+                flips++;
+                i = 1 - i;
+            } else {
+                throw new RuntimeException("Invalid color detected: " +
+                        Integer.toString(c, 16) + " instead of " + Integer.toString(fb[i], 16));
+            }
+        }
+        return flips;
+    }
+
+    private static void test(TextComponent tc) throws Exception {
+        if (tc instanceof TextField) {
+            System.out.println("TextField testing ...");
+        } else if (tc instanceof TextArea) {
+            System.out.println("TextArea testing ...");
+        }
+
+        // now passing along the component's vertical center, skipping 5px from both sides,
+        // we should see bg - textcolor - bg - selcolor - seltextcolor - selcolor - bg
+        // that is bg-fg-bg-fg-bg-fg-bg, 6 flips
+
+        EventQueue.invokeAndWait(() -> {
+            tc.setForeground(Color.green);
+            tc.setBackground(Color.magenta);
+
+            tc.setText("  I    I    ");
+            tc.select(5, 10);
+            tc.requestFocus();
+        });
+        r.waitForIdle();
+        r.delay(200);
+        EventQueue.invokeAndWait(() -> {
+            flips = countFlips(tc);
+        });
+        if (flips != 6) {
+            throw new RuntimeException("Invalid number of flips: " + flips + " instead of 6");
```

Comment:
Please wrap the long lines to a maximum of 80 characters per line.

Reply:
Updated.